module/zfs/ddt.c

```diff
@@ -719,6 +733,15 @@ ddt_lookup(ddt_t *ddt, const blkptr_t *bp, boolean_t add)
 
 	ddt_enter(ddt);
 
+	dde_temp = avl_find(&ddt->ddt_tree, &dde_search, &where);
+	if ((dde_temp == NULL) || (error == ENOENT && add == B_FALSE)) {
+		if (dde_temp != NULL)
+			avl_remove(&ddt->ddt_tree, dde);
+		dde->dde_loading = B_FALSE;
```

Comment:
do you need to `cv_broadcast(&dde->dde_cv);` after unsetting the loading flag? Trying to think through what would happen if there were two threads trying to load the same dde (one in `cv_wait()` on line 706) and then the DDE was removed (assuming the dedup ceiling threshold was crossed).

The cleanup here is hairy, it might be easier to mark the DDEs as "do not sync dde" and just ignore those DDEs during DDT sync.

Reply:
I think that's a good catch.  It hadn't been a problem before (nor have we run into it), but calling cv_broadcast shouldn't *hurt* anything.

The real problem with marking them as "do not sync" is that one of the big things sync does is clear out the avl cache; keeping some around in there is more complication than this.  (Or so I ended up thinking.)

Reply:
Ugh, maybe not.  It also calls ddt_free(), which is going to break anything that is using that specific dde.

Reply:
> The real problem with marking them as "do not sync" is that one of the big things sync does is clear out the avl cache; keeping some around in there is more complication than this. (Or so I ended up thinking.)

I don't believe you'd have to keep anything around in the avl cache. I think you could modify the existing sync mechanism to continue removing & destroying the DDEs while adding the ability to not sync some of them to disk.

Reply:
I have gone and looked at the original, untouched code again, and either the race condition I am worried about doesn't exist, or it has always existed and I have not introduced it.

Specifically, if ddt_sync_table() can run at the same time as ddt_lookup() (and I don't see any reason it coudn't, given that the big loop is outside of the ddt lock), then it could always end up calling ddt_free() on the entry in the AVL tree, and the code in ddt_lookup() would continue to use it.

Given that, I would prefer to go bak to the original, much-less-complicated code that made the same assumptions that the original code did.